### DISCUSSION TITLE
Portable RGB formats

### DISCUSSION THREAD

**Discussion by cryptoquick** [LNP-BP/LNPBPs#168] [Proposals]
It might be useful to standardize on a set of portable RGB formats for use between wallets. Sort of like how PDF is useful as a portable format for desktop publishing, we should standardize somewhat on some useful formats for if they're encoded and transmitted over copy and paste, or even automated Nostr DMs.

- RGB - strict-encoded binary
- RGBA - ascii-armored base85
- RGBAZ - ascii-armored base85, but compressed with xz
- RGBZ - strict-encoded binary, but compressed with xz

For a format like RGBAZ, a new field could be added, called Compression:

```
-----BEGIN RGB INTERFACE IMPLEMENTATION-----
Id: 9EUGHC-wpuiyrQE-NdPBVyiv-VX4sVRBs-9yKfteug-HtqnGb
Mnemonic: titanic-easy-citizen
IfaceId: 48hc4i-m9JRcYQA-uUSzwFCK-VNEa9eZf-nhepU8QJ-pqosXS
SchemaId: BEiLYE-am9WhTW1-oK8cpvw4-FEMtzMrf-mKocuGZn-qWK6YF
Compression: xz

0GJpk*8ec%bQl6%%><dOl0XO#=}srzy~EeF)&A22s4k2N#J#y%MwLq`h^$D7qg=O
qkP4LeiyAYvb}d8uJp|AP1aoj@V*mgE0MQ2oWMOn+(gzG_b8~fNWK(r;aBO
)12f2M1$vWnpw>WB>pF00E#63So0|Wpqz>Ze?-+$RB~Z%b7p07
-----END RGB INTERFACE IMPLEMENTATION-----
```

In this, only the base85 data is compressed.
Additional fields like ContractSource could also be nice for when Contractum is made and users wish to scrutinize the code that went into that contract.

Also, in case this is helpful information, I've found [lzma-rs](https://docs.rs/lzma-rs/latest/lzma_rs/) to be a good rust-native library for xz compression, which optimizes more towards size than speed, which is a better constraint in the scenarios of transferring a consignment or importing a contract.

**Comment by dr-orlovsky** [comment]
Agree with compression and standard extensions. One point: in RGBAZ we do not compress Base85 data; instead we compress the original data (i.e. take RGBZ) and encode them with Base85. Otherwise the performance and efficiency would be much lower (we would need to do double Base85 encoding + suffer lower compression efficiency).

**Reply by cryptoquick** [reply]
Correct, the flow should be:
strict_encoding (bytes) -> xz (bytes) -> base85 encoding (ascii)

**Comment by dr-orlovsky** [comment]
As for MediaTypes (MIMES) I was thinking about

- `application/lnp-bp.rgb` for *.rgb
- `application/lnp-bp.rgb+zip` for *.rgbz
- `application/lnp-bp.rgb; encoding=base85` for *.rgba
- `application/lnp-bp.rgb+zip; encoding=base85` for *.rgbaz

PS. All data types - interfaces, contracts, schemata etc are the same encoding, thanks to Bindle system, with header specifying what's inside.

**Comment by dr-orlovsky** [comment]
Also wanted to ask why specifically xz compression and not something more widespread?

**Reply by cryptoquick** [reply]
It's the best of the type of algorithms used in gzip without going to zstd. The problem with zstd is, it's not written in native Rust, so I don't know how portable it would be. But xz is superior to pretty much any other algorithm in terms of compression size, though it does sacrifice some in performance. In this case, for messaging encoding, I think that's fine.

**Reply by dr-orlovsky** [reply]
What about `inflate` crate? Its native, I used it someday in RGB <v0.9. I thought it implements standard for zip.

**Reply by cryptoquick** [reply]
I saw that, which is what inspired me to want compression in the first place. Inflate doesn't include support for xz.
